core/lib/multivm/src/versions/vm_fast/tracers/storage.rs

```diff
@@ -0,0 +1,46 @@
+use zksync_vm2::interface::{GlobalStateInterface, Opcode, OpcodeType, ShouldStop, Tracer};
+
+use crate::interface::storage::{StoragePtr, WriteStorage};
+
+/// Tracer that stops VM execution once too many storage slots were read from the underlying storage.
+/// Useful as an anti-DoS measure.
+#[derive(Debug)]
+pub struct StorageInvocationsTracer<ST> {
+    // FIXME: this is very awkward; logically, storage should be obtainable from `GlobalStateInterface` (but how?)
```

Comment:
Subj. FWIW, during local testing, I've observed no measurable performance difference between a VM run with and without this tracer (see the changes in the `batch` VM benchmark).

Reply:
There should be no effect because storage access is expensive anyway.